dwst/scripts/particles.js

```diff
@@ -20,7 +20,32 @@ const specialChars = [
   '\\',
 ];
 
-export class InvalidParticles extends Error { }
+export class InvalidParticles extends Error {
+  constructor(expression, expected, expectedStringLength, remainder) {
```

Comment:
Since you often pass null in as `expression`, you should probably turn `expression` into an optional argument for the constructor. Just write `expression=null` and make it the last argument of the constructor. You can then simply omit the null when making the call.

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters

Reply:
I'm not 100% sure if default parameters work in the constructor but at least you could give it a try to see if that works.

Reply:
could it also be just left undefined?

Reply:
```
function foo(bar) {
  // bar is undefined but it might be a pogramming error
  console.log(bar);
}
foo()
```

```
function foo(bar) {
  // bar is undefined and if you see the call you know it is intentional
  console.log(bar);
}
foo(undefined)
```

```
function foo(bar) {
  // bar is null and you know it is intentional without looking
  console.log(bar);
}
foo(null)
```

Reply:
Those are examples about *implicit* `undefined` vs *explicit* `undefined` vs `null`

Reply:
Then there is the question about default parameters

Reply:
```
function foo(bar) {
  // passing bar might be required
}
```

```
function foo(bar=undefined) {
  // passing bar is clearly optional
}
```

Reply:
And therefore I recommend you do

Reply:
```
function foo(bar=null) {
  // ...
}
```

Reply:
Nope. Everything broke.